File: experiments/codex-oauth-gateway/src/codex/response.ts
import type { GatewayRequestBody } from "../types.js";

export function wantsStream(body: GatewayRequestBody): boolean {
	return body.stream === true;
}

export async function convertSseToJson(response: Response): Promise<Response> {
	if (!response.body) {
		return response;
	}

	const text = await response.text();
	const finalResponse = parseFinalResponse(text);
	const headers = ensureContentType(response.headers);
	headers.set("content-type", "application/json; charset=utf-8");

	if (!finalResponse) {
		return new Response(text, {
			status: response.status,
			statusText: response.statusText,
			headers: ensureContentType(response.headers),
		});
	}

	return new Response(JSON.stringify(finalResponse), {
		status: response.status,
		statusText: response.statusText,
		headers,
	});
}

export function ensureContentType(headers: Headers): Headers {
	const responseHeaders = new Headers(headers);

	if (!responseHeaders.has("content-type")) {
		responseHeaders.set("content-type", "text/event-stream; charset=utf-8");
	}

	return responseHeaders;
}

export async function handleErrorResponse(response: Response): Promise<Response> {
	const mapped = await mapUsageLimit404(response);
	return mapped ?? response;
}

export async function handleSuccessResponse(
	response: Response,
	isStreaming: boolean,
): Promise<Response> {
	const responseHeaders = ensureContentType(response.headers);

	if (!isStreaming) {
		return await convertSseToJson(
			new Response(response.body, {
				status: response.status,
				statusText: response.statusText,
				headers: responseHeaders,
			}),
		);
	}

	return new Response(response.body, {
		status: response.status,
		statusText: response.statusText,
		headers: responseHeaders,
	});
}

async function mapUsageLimit404(response: Response): Promise<Response | null> {
	if (response.status !== 404) return null;

	const clone = response.clone();
	let text = "";
	try {
		text = await clone.text();
	} catch {
		text = "";
	}
	if (!text) return null;

	let code = "";
	try {
		const parsed = JSON.parse(text) as {
			error?: { code?: string; type?: string };
		};
		code = (parsed?.error?.code ?? parsed?.error?.type ?? "").toString();
	} catch {
		code = "";
	}

	if (code !== "usage_limit_exceeded") {
		return null;
	}

	return new Response(response.body, {
		status: 429,
		statusText: "Too Many Requests",
		headers: response.headers,
	});
}

function parseFinalResponse(sseText: string): unknown | null {
	for (const line of sseText.split("\n")) {
		if (!line.startsWith("data: ")) continue;
		try {
			const event = JSON.parse(line.slice(6)) as {
				type?: string;
				response?: unknown;
			};
			if (
				event.type === "response.done" ||
				event.type === "response.completed"
			) {
				return event.response ?? null;
			}
		} catch {
			// Ignore malformed SSE data lines.
		}
	}
	return null;
}

File: codex-oauth-gateway-python/gateway/response.py
from __future__ import annotations

import json


def parse_final_response(sse_text: str):
    text_deltas: list[str] = []
    output_items: list[dict] = []
    output_text_done: str | None = None
    final_response = None

    for line in sse_text.splitlines():
        if not line.startswith("data: "):
            continue
        try:
            event = json.loads(line[6:])
        except Exception:
            continue

        if event.get("type") in {"response.output_text.delta", "output_text.delta"}:
            delta = event.get("delta")
            if isinstance(delta, str):
                text_deltas.append(delta)
            continue

        if event.get("type") == "response.output_text.done":
            done_text = event.get("text")
            if isinstance(done_text, str):
                output_text_done = done_text
            continue

        if event.get("type") == "response.output_item.done":
            item = event.get("item")
            if isinstance(item, dict):
                output_items.append(item)
            continue

        if event.get("type") in {"response.done", "response.completed"}:
            final_response = event.get("response")

    if not final_response:
        return None

    if not isinstance(final_response, dict):
        return final_response

    output = final_response.get("output")
    has_output = isinstance(output, list) and len(output) > 0
    if has_output:
        return final_response

    if output_items:
        final_response["output"] = output_items
        if not final_response.get("output_text"):
            text_parts: list[str] = []
            for item in output_items:
                if item.get("type") != "message":
                    continue
                for part in item.get("content") or []:
                    if isinstance(part, dict) and part.get("type") == "output_text" and isinstance(part.get("text"), str):
                        text_parts.append(part["text"])
            if text_parts:
                final_response["output_text"] = "".join(text_parts)
        return final_response

    combined_text = output_text_done if output_text_done is not None else "".join(text_deltas)
    if combined_text:
        final_response["output_text"] = combined_text
        final_response["output"] = [
            {
                "type": "message",
                "role": "assistant",
                "content": [{"type": "output_text", "text": combined_text}],
            }
        ]

    return final_response


def map_usage_limit_404(status_code: int, body_text: str) -> tuple[int, str]:
    if status_code != 404:
        return status_code, body_text
    try:
        parsed = json.loads(body_text)
        code = (parsed.get("error") or {}).get("code") or (parsed.get("error") or {}).get("type")
        if code == "usage_limit_exceeded":
            return 429, body_text
    except Exception:
        pass
    return status_code, body_text
